geo/geogfn: implement geography intersects using S2

otan opened the pull request:
This commits adds the intersect geography functionality, which uses the
S2 library to do the heavy lifting. Some of the edge crossings we've had
to implement ourselves but should be straightforward.

We have a 1cm precision for this (PostGIS is 0.001cm) which is line with
the S2 library, but that should be good enough for geospatial use cases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47791)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47791)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/intersects.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Xx0gw6qsGYeSM_WZz:-M5Xx0gw6qsGYeSM_W_-:b-4wo89y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5100df01d8a92235373028f0e1fff5ca520dfcf4/pkg/geo/geogfn/intersects.go#L93)):*
> ```Go
> 		edge := a.Edge(aEdgeIdx)
> 		crosser := s2.NewChainEdgeCrosser(edge.V0, edge.V1, (*b)[0])
> 		for _, nextVertex := range (*b)[1:] {
> ```

I am not sure about the semantics of `s2.Crossing` -- what happens if one edge is contained inside the other? The endpoints (vertices in s2 terminology?) don't have to match.
___
*[pkg/geo/geogfn/intersects.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5XxjrB-YURPDnkea0e:-M5XxjrB-YURPDnkea0f:b-zi5jz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5100df01d8a92235373028f0e1fff5ca520dfcf4/pkg/geo/geogfn/intersects.go#L112)):*
> ```Go
> 			crosser := s2.NewChainEdgeCrosser(loopEdge.V0, loopEdge.V1, (*b)[0])
> 			for _, nextVertex := range (*b)[1:] {
> 				crossing := crosser.ChainCrossingSign(nextVertex)
> ```

a polyline contained inside a polygon is also intersecting -- is that handled here?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47791)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/intersects.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Xx0gw6qsGYeSM_WZz:-M5YBTe58KLv8RnoDNWw:bsv5u9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5100df01d8a92235373028f0e1fff5ca520dfcf4/pkg/geo/geogfn/intersects.go#L93)):*
I have a test for this. Seems to work.
> CrossingSign reports whether the edge AB intersects the edge CD.
so yeah!
___
*[pkg/geo/geogfn/intersects.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5XxjrB-YURPDnkea0e:-M5YTV8xD4BtkqHnOrGV:b-k2rwtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5100df01d8a92235373028f0e1fff5ca520dfcf4/pkg/geo/geogfn/intersects.go#L112)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

a polyline contained inside a polygon is also intersecting -- is that handled here?
</blockquote></details>

Good catch. Fixed.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47791)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/intersects.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Ym33g0M14bOr3P9o1:-M5Ym33g0M14bOr3P9o2:bior2h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects.go#L106)):*
> ```Go
> // polyline b.
> func polygonIntersectsPolyline(a *s2.Polygon, b *s2.Polyline) bool {
> 	// Check if the polygon contains any point of the line b.
> ```

s/point/vertex/
___
*[pkg/geo/geogfn/intersects.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Ym5Yh1Telx3JSb5Zp:-M5Ym5Yh1Telx3JSb5Zq:b-ttxmtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects.go#L111)):*
> ```Go
> 			return true
> 		}
> 	}
> ```

Can you add a comment like:
The polygon does not contain any vertex of the polyline. The polyline can intersect the polygon if-and-only if a line in the polyline crosses from outside the polygon to inside the polygon or vice versa.
- For a polygon with no holes, this is true if any edge of the polygon and any line of the polyline cross since the line must be crossing from inside to outside or vice versa.
- For a polgon with holes ...

Hmm, I am not sure about the latter. Can the holes touch the outer loop and can the holes touch each other? And if yes, is the number of points where they touch allowed to be infinite? And are the points on the boundary of the hole considered inside or outside the polygon? Does that answer change if the point is on the boundary of both the outer loop and the hole?
I am trying to see if a line in the polyline can pass between two touching holes without being inside the polygon. Also is all of this irrelevant since we have 1cm precision, so any touching holes will not really be touching and one can consider that the leaf cell corresponding to that touching is also part of the polygon?

It would be nice if we could define the (achievable, given what we have in S2) semantics in a crisper manner, how it differs from PostGIS, and then use it to guide our reasoning of correctness of our code.
___
*[pkg/geo/geogfn/intersects_test.go, line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5YkooW1eofI7gawwbY:-M5YkooW1eofI7gawwbZ:bs82p0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects_test.go#L98)):*
> ```Go
> 			"LINESTRING(1.5499860 1.5501575, 3.0 3.0)",
> 			true,
> 		},
> ```

can you add a test of the form
LINESTRING(1 1 , 4 4)
LINESTRING(2 2, 3 3)

which should return true (I didn't see a test where one line string is completely inside the other).


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47791)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/intersects.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Ym33g0M14bOr3P9o1:-M5Yu39b-yT9ga4qwBnu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects.go#L106)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/point/vertex/
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/intersects.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Ym5Yh1Telx3JSb5Zp:-M5YuhXb8J9nAlLzx5DB:bcgfqf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects.go#L111)):*
I've added a comment to that effect. Not sure holes matter here.

> I am trying to see if a line in the polyline can pass between two touching holes without being inside the polygon. Also is all of this irrelevant since we have 1cm precision, so any touching holes will not really be touching and one can consider that the leaf cell corresponding to that touching is also part of the polygon?

It will intersect at the point the holes touch. Consider this polygon with two holes which touch at 0.5 0.3.

Source of truth: GEOS

```

otan=# select st_intersects('POLYGON((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0), (0.3 0.3, 0.4 0.2, 0.5 0.3, 0.4 0.4, 0.3 0.3), (0.5 0.3, 0.6 0.2, 0.7 0.3, 0.6 0.4, 0.5 0.3))', 'LINESTRING(0.4 0.3, 0.6 0.3)');
 st_intersects
---------------
 t
(1 row)
```
___
*[pkg/geo/geogfn/intersects_test.go, line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5YkooW1eofI7gawwbY:-M5YyHKDBeSJZqjGyH1S:b-6nwkoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects_test.go#L98)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add a test of the form
LINESTRING(1 1 , 4 4)
LINESTRING(2 2, 3 3)

which should return true (I didn't see a test where one line string is completely inside the other).
</blockquote></details>

Done, but slightly different as the lines as you described are not on top of each other in a geodesic sense.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-:-M5ZeBXxA0scU7dIcy9L:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/intersects.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47791#-M5Ym5Yh1Telx3JSb5Zp:-M5ZdPM8DJcZWPH2rNw1:bnnyzzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1761acca2a2cf62573cebd1d645aab54191dbe35/pkg/geo/geogfn/intersects.go#L111)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I've added a comment to that effect. Not sure holes matter here.

> I am trying to see if a line in the polyline can pass between two touching holes without being inside the polygon. Also is all of this irrelevant since we have 1cm precision, so any touching holes will not really be touching and one can consider that the leaf cell corresponding to that touching is also part of the polygon?

It will intersect at the point the holes touch. Consider this polygon with two holes which touch at 0.5 0.3.

Source of truth: GEOS

```

otan=# select st_intersects('POLYGON((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0), (0.3 0.3, 0.4 0.2, 0.5 0.3, 0.4 0.4, 0.3 0.3), (0.5 0.3, 0.6 0.2, 0.7 0.3, 0.6 0.4, 0.5 0.3))', 'LINESTRING(0.4 0.3, 0.6 0.3)');
 st_intersects
---------------
 t
(1 row)
```
</blockquote></details>

Can you add a code comment with this example and saying that this approach works even for holes that are touching and holes that touch the exterior (I am assuming the same applies to the latter).


<!-- Sent from Reviewable.io -->

otan committed:
geo/geogfn: implement geography intersects using S2

This commits adds the intersect geography functionality, which uses the
S2 library to do the heavy lifting. Some of the edge crossings we've had
to implement ourselves but should be straightforward.

We have a 1cm precision for this (PostGIS is 0.001cm) which is line with
the S2 library, but that should be good enough for geospatial use cases.

Release note: None
